pkg/scheduler/plugins/gpushare/device_cache.go

```diff
@@ -0,0 +1,132 @@
+	// and uses node name as map key.
+	nodeDeviceInfos map[string]*nodeDevice
+}
+
+type nodeDevice struct {
+	lock        sync.Mutex
+	deviceTotal map[schedulingv1alpha1.DeviceType]*deviceResources
+	deviceFree  map[schedulingv1alpha1.DeviceType]*deviceResources
+	deviceUsed  map[schedulingv1alpha1.DeviceType]*deviceResources
+	allocateSet map[string]*framework.PodInfo
+}
+
+// deviceResources is used to present resources per device.
+// we use the minor of device as key
+// "0": {koordinator.sh/gpu-core:100, koordinator.sh/gpu-memory-ratio:100, koordinator.sh/gpu-memory: 16GB}
+// "1": {koordinator.sh/gpu-core:100, koordinator.sh/gpu-memory-ratio:100, koordinator.sh/gpu-memory: 16GB}
+type deviceResources map[int]corev1.ResourceList
+
+func newNodeDeviceCache() *nodeDeviceCache {
+	return &nodeDeviceCache{
+		nodeDeviceInfos: make(map[string]*nodeDevice),
+	}
+}
+
+func newNodeDevice() *nodeDevice {
+	return &nodeDevice{
+		deviceTotal: make(map[schedulingv1alpha1.DeviceType]*deviceResources),
+		deviceFree:  make(map[schedulingv1alpha1.DeviceType]*deviceResources),
+		deviceUsed:  make(map[schedulingv1alpha1.DeviceType]*deviceResources),
+		allocateSet: make(map[string]*framework.PodInfo),
+	}
+}
+
+func (n *nodeDeviceCache) getNodeDevice(nodeName string) *nodeDevice {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	return n.nodeDeviceInfos[nodeName]
+}
+
+func (n *nodeDevice) resetDeviceFree(dType schedulingv1alpha1.DeviceType) {
+	for minor, usedResource := range *n.deviceUsed[dType] {
+		(*n.deviceFree[dType])[minor] = quotav1.SubtractWithNonNegativeResult(
+			(*n.deviceTotal[schedulingv1alpha1.GPU])[minor],
+			usedResource)
+	}
+}
+
+func (n *nodeDevice) updateDeviceUsed(dType schedulingv1alpha1.DeviceType, allocation *apiext.DeviceAllocation, add bool) {
+	if add {
+		(*n.deviceUsed[dType])[int(allocation.Minor)] = quotav1.Add(
+			(*n.deviceUsed[dType])[int(allocation.Minor)],
+			allocation.Resources)
+	} else {
+		(*n.deviceUsed[dType])[int(allocation.Minor)] = quotav1.SubtractWithNonNegativeResult(
+			(*n.deviceUsed[dType])[int(allocation.Minor)],
+			allocation.Resources)
+	}
+}
+
+func (n *nodeDevice) tryAllocateGPU(podRequest corev1.ResourceList) []*apiext.DeviceAllocation {
+	var allocateResult []*apiext.DeviceAllocation
+	if isMultipleGPUPod(podRequest) {
+		gpuCore, gpuMem, gpuMemRatio := podRequest[apiext.GPUCore], podRequest[apiext.GPUMemory], podRequest[apiext.GPUMemoryRatio]
+		gpuWanted := gpuCore.Value() % 100
+		podRequestPerCard := corev1.ResourceList{
+			apiext.GPUCore:        *resource.NewQuantity(gpuCore.Value()/gpuWanted, resource.DecimalSI),
+			apiext.GPUMemory:      *resource.NewQuantity(gpuMem.Value()/gpuWanted, resource.BinarySI),
+			apiext.GPUMemoryRatio: *resource.NewQuantity(gpuMemRatio.Value()/gpuWanted, resource.DecimalSI),
+		}
+		satisfiedDeviceCount := 0
+		for minor, resources := range *n.deviceFree[schedulingv1alpha1.GPU] {
+			if satisfiedDeviceCount == int(gpuWanted) {
```

Comment:
if the last minor make satisfiedDeviceCount == int(gpuWanted)  you have logic bug that can't return allocateResult correctly.  "if satisfiedDeviceCount == int(gpuWanted) return allocateResult" should be execute out of "for", and here should be "if satisfiedDeviceCount == int(gpuWanted) break"

Reply:
Thanks for pointing that out! I moved `satisfiedDeviceCount == int(gpuWanted)` after `quotav1.LessThanOrEqual`, that should work.